# **Automatidata project**
**Course 2 - Get Started with Python**

Welcome to the Automatidata Project!

You have just started as a data professional in a fictional data consulting firm, Automatidata. Their client, the New York City Taxi and Limousine Commission (New York City TLC), has hired the Automatidata team for its reputation in helping their clients develop data-based solutions.

The team is still in the early stages of the project. Previously, you were asked to complete a project proposal by your supervisor, DeShawn Washington. You have received notice that your project proposal has been approved and that New York City TLC has given the Automatidata team access to their data. To get clear insights, New York TLC's data must be analyzed, key variables identified, and the dataset ensured it is ready for analysis.

A notebook was structured and prepared to help you in this project. Please complete the following questions.

# Course 2 End-of-course project: Inspect and analyze data

In this activity, you will examine data provided and prepare it for analysis.  This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>    

**The purpose** of this project is to investigate and understand the data provided.
  
**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings. 
<br/>  
*This activity has three parts:*

**Part 1:** Understand the situation 
* Prepare to understand and organize the provided taxi cab dataset and information.

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities.

* Compile summary information about the data to inform next steps.

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into specific variables.


<br/> 
Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work. 



# **Identify data types and relevant variables using Python**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## PACE: **Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided taxi cab information? 

Understand what is the goal of the project and gather information about the data that is provided.

<img src="images/Analyze.png" width="100" height="100" align=left>

## PACE: **Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Build dataframe**
















Create a pandas dataframe for data learning, and future exploratory data analysis (EDA) and statistical activities.

**Code the following,**

*   import pandas as `pd`. pandas is used for buidling dataframes.

*   import numpy as `np`. numpy is imported with pandas

*   `df = pd.read_csv('Datasets\NYC taxi data.csv')`

**Note:** pair the data object name `df` with pandas functions to manipulate data, such as `df.groupby()`.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [1]:
#Import libraries and packages listed above
import pandas as pd
import numpy as np

# Load dataset into dataframe
df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
print("done")

done


### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by coding the following:

1. `df.head(10)`
2. `df.info()`
3. `df.describe()`

Consider the following two questions:

**Question 1:** When reviewing the `df.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 2:** When reviewing the `df.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values?

**Question 1 answer:**
The first column is missing the column name. Looking at all columns contained in the loaded dataframe and comparing them to the column names provided in the description of the data, the first column contains the ID.

There is a mix of naming conventions used for the column names: some columns use camel case while others use snake case.

There are no null values in the dataframe.

Most columns contain numerical data (8 contain floats and 7 contain ints). 3 columns contain other types of data.


**Question 2 answer:**
The columns that contain ints correspond to ids or codes, with the exception of the passenger_count which is a count. These columns contain a restricted number of values.

Some of the columns that contains values related to fares contain negative values, which are questionable.

In [2]:
df.head(10)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   
5    23345809         2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
6    37660487         2   05/03/2017 7:04:09 PM   05/03/2017 8:03:47 PM   
7    69059411         2   08/15/2017 5:41:06 PM   08/15/2017 6:03:05 PM   
8     8433159         2   02/04/2017 4:17:07 PM   02/04/2017 4:29:14 PM   
9    95294817         1   11/10/2017 3:20:29 PM   11/10/2017 3:40:55 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   
5                6           2.30           1                  N   
6                1          12.83           1                  N   
7                1           2.98           1                  N   
8                1           1.20           1                  N   
9                1           1.60           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   
5           161           236             1          9.0    0.5      0.5   
6            79           241             1         47.5    1.0      0.5   
7           237           114             1         16.0    1.0      0.5   
8           234           249             2          9.0    0.0      0.5   
9           239           237             1         13.0    0.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80  
5        2.06           0.0                    0.3         12.36  
6        9.86           0.0                    0.3         59.16  
7        1.78           0.0                    0.3         19.58  
8        0.00           0.0                    0.3          9.80  
9        2.75           0.0                    0.3         16.55

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [4]:
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

In [5]:
len(set(df["Unnamed: 0"])) == df.shape[0]

True

In [6]:
df["extra"].sort_values(ascending=False)[:20]

10722    4.5
13947    4.5
6477     4.5
5862     4.5
10032    4.5
18754    4.5
20251    4.5
10044    4.5
7867     4.5
3920     4.5
1371     4.5
15148    4.5
6227     4.5
2726     4.5
21148    4.5
18729    4.5
18954    4.5
10539    4.5
8468     4.5
1927     4.5
Name: extra, dtype: float64

### **Task 2c. Understand the data - Investigate the variables**

Sort and interpret the data table for two variables:`trip_distance` and `total_amount`.

**Answer the following three questions:**

**Question 1:** Sort your first variable (`trip_distance`) from maximum to minimum value, do the values seem normal?

**Question 2:** Sort by your second variable (`total_amount`), are any values unusual?

**Question 3:** Are the resulting rows similar for both sorts? Why or why not?

In [7]:
df["trip_distance"].sort_values(ascending=False)

# Sort the data by trip distance from maximum to minimum value

9280     33.96
13861    33.92
6064     32.72
10291    31.95
29       30.83
         ...  
2440      0.00
15916     0.00
1350      0.00
246       0.00
17788     0.00
Name: trip_distance, Length: 22699, dtype: float64

In [8]:
(df["trip_distance"] == 0).sum()

148

In [9]:
df.loc[df["trip_distance"] == 0, "total_amount"][:20]

128     20.30
246      3.80
291      4.30
319      3.80
424     10.00
470     41.76
472     11.30
572     70.27
647      6.80
795      8.80
920      3.80
1205     4.30
1275     3.80
1350     3.96
1449    63.00
1517    70.06
1638    52.39
1804    14.04
2150     4.75
2440     6.80
Name: total_amount, dtype: float64

There are 148 records for which the trip distance is 0, which could be an error given that the total fare amount for many of these trips is quite high.

In [10]:
df["total_amount"].sort_values(ascending=False)[:20]

# Sort the data by total amount and print the top 20 values

8476     1200.29
20312     450.30
13861     258.21
12511     233.74
15474     211.80
6064      179.06
16379     157.06
3582      152.30
11269     151.82
9280      150.30
1928      137.80
10291     131.80
6708      126.00
11608     123.30
908       121.56
7281      120.96
18130     119.31
13621     115.94
13359     111.95
29        111.38
Name: total_amount, dtype: float64

In [11]:
df["total_amount"].sort_values(ascending=False)[-20:]

# Sort the data by total amount and print the bottom 20 values

14283      0.31
19067      0.30
10506      0.00
5722       0.00
4402       0.00
22566      0.00
1646      -3.30
18565     -3.80
314       -3.80
5758      -3.80
5448      -4.30
4423      -4.30
10281     -4.30
8204      -4.80
20317     -4.80
11204     -5.30
14714     -5.30
17602     -5.80
20698     -5.80
12944   -120.30
Name: total_amount, dtype: float64

In [12]:
df[["trip_distance", "total_amount"]].sort_values(by="total_amount", ascending=False)[:20]

trip_distance  total_amount
8476            2.60       1200.29
20312           0.00        450.30
13861          33.92        258.21
12511           0.00        233.74
15474           0.00        211.80
6064           32.72        179.06
16379          25.50        157.06
3582            7.30        152.30
11269           0.00        151.82
9280           33.96        150.30
1928           12.50        137.80
10291          31.95        131.80
6708            0.32        126.00
11608          23.00        123.30
908            26.12        121.56
7281            0.00        120.96
18130          30.50        119.31
13621          19.80        115.94
13359           0.00        111.95
29             30.83        111.38

In [13]:
df[["trip_distance", "total_amount"]].sort_values(by="total_amount", ascending=False)[-20:]

trip_distance  total_amount
14283           0.00          0.31
19067           0.10          0.30
10506           0.00          0.00
5722            0.00          0.00
4402            7.06          0.00
22566           0.00          0.00
1646            0.04         -3.30
18565           0.10         -3.80
314             0.12         -3.80
5758            0.02         -3.80
5448            0.25         -4.30
4423            0.06         -4.30
10281           0.00         -4.30
8204            0.41         -4.80
20317           0.24         -4.80
11204           0.64         -5.30
14714           0.40         -5.30
17602           0.46         -5.80
20698           0.70         -5.80
12944           0.17       -120.30

In [14]:
df[["fare_amount", "total_amount"]].sort_values(by="total_amount", ascending=False)[:20]

fare_amount  total_amount
8476        999.99       1200.29
20312       450.00        450.30
13861       200.01        258.21
12511       175.00        233.74
15474       200.00        211.80
6064        107.00        179.06
16379       140.00        157.06
3582        152.00        152.30
11269       120.00        151.82
9280        150.00        150.30
1928        120.00        137.80
10291       131.00        131.80
6708        100.00        126.00
11608        99.50        123.30
908         100.00        121.56
7281        100.00        120.96
18130        90.50        119.31
13621       105.00        115.94
13359        75.00        111.95
29           80.00        111.38

In [15]:
(df["fare_amount"] < 0).sum()

14

In [16]:
(df["fare_amount"] == 0).sum()

6

**Answers to questions 1 to 3:**

There are some pretty high values for total amounts, which on their own could be realistic. However, when looking at the trip distance for those values, several of them have a distance of 0, or a short one, which could not justify the high fare amounts.

Additionally, there are some records that have a total fare amount that is less than or equal to 0, which also seem unrealistic.

Some of the extreme fare amounts are for records where the trip distance is 0. Therefore there is an overlap between some of the abnormal values for these two variables, but only in a minority of cases.

For the both positive and negative extreme values for the fare amounts, it would be worth asking Titus from the TLC team whether such fare amounts are possible under certain circumstances. In case of doubt, these records can be removed before conducting statistical analyses.

In [17]:
df["payment_type"].value_counts()

# How many of each payment type are represented in the data?

1    15265
2     7267
3      121
4       46
Name: payment_type, dtype: int64

There are no instances of payment types 5 and 6.

According to the data dictionary, the payment method was encoded as follows:

1 = Credit card  
2 = Cash  
3 = No charge  
4 = Dispute  
5 = Unknown  
6 = Voided trip

In [18]:
df.loc[df["payment_type"] == 1, "tip_amount"].mean()

# What is the average tip for trips paid for with credit card?

2.7298001965279934

In [19]:
df.loc[df["payment_type"] == 2, "tip_amount"].mean()

# What is the average tip for trips paid for with cash?

0.0

In [20]:
(df.loc[df["payment_type"] == 2, "tip_amount"] == 0).all()

True

In [21]:
numerical_cols = [
    "passenger_count", 
    "trip_distance", 
    "fare_amount", 
    "extra", "mta_tax",
    "tip_amount",
    "tolls_amount",
    "improvement_surcharge",
    "total_amount"
]
df.groupby("payment_type")[numerical_cols].mean()

passenger_count  trip_distance  fare_amount     extra   mta_tax  \
payment_type                                                                    
1                    1.622142       3.037661    13.429748  0.340485  0.498133   
2                    1.692583       2.647486    12.213546  0.319527  0.498349   
3                    1.305785       2.986612    12.186116  0.326446  0.413223   
4                    1.282609       3.450652     9.913043  0.130435  0.347826   

              tip_amount  tolls_amount  improvement_surcharge  total_amount  
payment_type                                                                 
1                 2.7298      0.357659               0.299902     17.663577  
2                 0.0000      0.214441               0.299959     13.545821  
3                 0.0000      0.388595               0.265289     13.579669  
4                 0.0000      0.638261               0.208696     11.238261

In [22]:
df["VendorID"].value_counts()

# How many times is each vendor ID represented in the data?

2    12626
1    10073
Name: VendorID, dtype: int64

In [23]:
df.groupby("VendorID")["total_amount"].mean()

# What is the mean total amount for each vendor?

VendorID
1    16.298119
2    16.320382
Name: total_amount, dtype: float64

In [24]:
df.groupby("VendorID")[numerical_cols].agg(["mean", "min", "max"])

passenger_count         trip_distance             fare_amount         \
                    mean min max          mean  min    max        mean    min   
VendorID                                                                        
1               1.258910   0   6      2.881485  0.0  30.50   13.035511    0.0   
2               1.948202   1   6      2.938705  0.0  33.96   13.019544 -120.0   

                     extra  ... tip_amount tolls_amount              \
             max      mean  ...        max         mean  min    max   
VendorID                    ...                                       
1         999.99  0.332423  ...      200.0     0.295119  0.0  18.28   
2         450.00  0.333954  ...       55.5     0.326441  0.0  19.10   

         improvement_surcharge           total_amount                  
                          mean  min  max         mean    min      max  
VendorID                                                               
1                     0.300000  0.3  0.3    16.298119    0.3  1200.29  
2                     0.299192 -0.3  0.3    16.320382 -120.3   450.30  

[2 rows x 27 columns]

In [25]:
df_cc = df[df["payment_type"] == 1]

# Filter the data for credit card payments only

df_cc["passenger_count"]

# Filter the credit-card-only data for passenger count only

0        6
1        1
2        1
3        1
5        6
        ..
22692    1
22693    1
22695    1
22697    1
22698    1
Name: passenger_count, Length: 15265, dtype: int64

In [26]:
df_cc.groupby("passenger_count")["tip_amount"].mean()

# Calculate the average tip amount for each passenger count (credit card payments only)


passenger_count
0    2.610370
1    2.714681
2    2.829949
3    2.726800
4    2.607753
5    2.762645
6    2.643326
Name: tip_amount, dtype: float64

In [27]:
df_cc.groupby("passenger_count")["fare_amount"].mean()

passenger_count
0    15.544444
1    13.278924
2    14.291813
3    13.431667
4    13.655431
5    13.154839
6    13.166297
Name: fare_amount, dtype: float64

In [28]:
df.groupby("passenger_count")["fare_amount"].mean()

passenger_count
0    14.157576
1    12.793423
2    13.984614
3    13.486359
4    13.683538
5    12.931759
6    12.920635
Name: fare_amount, dtype: float64

In [29]:
cols_to_sum = [
    "fare_amount",
    "extra",
    "mta_tax",
    "tip_amount",
    "tolls_amount",
    "improvement_surcharge"
]
(df[cols_to_sum].sum(axis=1) == df["total_amount"]).sum()/len(df)

0.8408740473148597

In [30]:
extra_charges = [
    "extra",
    "mta_tax",
    "tolls_amount",
    "improvement_surcharge"
]

<img src="images/Construct.png" width="100" height="100" align=left>

## PACE: **Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project. 




<img src="images/Execute.png" width="100" height="100" align=left>

## PACE: **Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response.


### **Given your efforts, what can you summarize for DeShawn and the data team?**

*Note for Learners: Your notebook should contain data that can address Luana's requests. Which two variables are most helpful for building a predictive model for the client: NYC TLC?*

The dataframe received contains 18 columns, of which 8 contain float values, 7 contain int values, 2 contain datetime objects and 1 contains strings.

There are no missing values in this dataset.

I identified the following columns as not being relevent for this project:
- VendorID: the mean is very similar for both vendors, and there is no reason to assume that the type of vendor as an influence on the fare amount
- Store_and_fwd_flag: this variable stores a technical aspect regarding the trip record, which as no bearing on the fare amount
- All fare-related variables apart from fare_amount: if we only want to estimate the fare amount, all other variables that contribute to the total amount are irrelevent for this project. In particular the tip amount, since that should not be included in the fare estimation provided to customers

We should exclude the records for payment types other than 1 and 2 (credit cards and cash) as they are few and the data they contain might not be useful for estimating the fare amount.

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.